### ISSUE TITLE
Deprecating the DigestMethod entry in SigRef dictionaries

### ISSUE THREAD

**Issue by MatthiasValvekens** [pdf-association/pdf-issues#117]
## Describe the bug

The **DigestMethod** key in a signature reference dictionary is defined as follows in ISO 32000-2:2020, Table 256:

> (Required) A name identifying the algorithm that shall be used when computing the digest if not specified in the certificate. Valid values are _MD5, SHA1 SHA256, SHA384_, _SHA512_ and _RIPEMD160_.
> NOTE (2020) The use of MD5 and SHA1 are deprecated with PDF 2.0. The **DigestMethod** key was also corrected to be required as no default value is defined.
>
>  Default value for PDF 1.5-1.7: _MD5_.

My (possibly mistaken) understanding is that this entry is a holdover from the pre-ISO PDF 1.7 specification, where it was used together with the **DigestValue** and **DigestLocation** entries to compute a digest over the portion of the PDF document indicated by the **Data** entry, and that this feature was removed from the ISO spec because of interoperability concerns. **DigestValue** and **DigestLocation** were removed in the ISO version of the specification, but **DigestMethod** was kept around. (@lrosenthol I hope I'm not mischaracterising the historical context)

Now, as for the current definition: from the point of view of verifying a modern CMS-based signature (i.e. `adbe.pkcs7.detached` , `ETSI.CAdES.detached` or even `adbe.pkcs7.sha1`), taking into account the fact that all digests are computed over a **ByteRange** these days, the definition in ISO 32000-2:2020 doesn't make a lot of sense.

 - In a CMS or PKCS#7 context, the algorithm used to digest the document is not specified in the signer's certificate, but in the `digestAlgorithm` field in the relevant `signerInfo` value in the CMS object. 
     - This is a [required field](https://datatracker.ietf.org/doc/html/rfc5652#section-5.3), so it can't be omitted without making the signature effectively unparseable.
     - There's no reason to suppose that the digest used in the CA's signature on the signer's certificate would be the same as the one used to sign the document. They're signatures by different entities for different purposes; they can use completely unrelated signature algorithms, and completely unrelated digest functions too.
 - Since the **DigestMethod** entry is defined as part of a signature _reference_ dictionary, and not the signature dictionary, I find it very hard to believe that its value could ever have had anything to do with something other than an object digest computed over the  **Data** entry.

**TL;DR**: I can't see what purpose this entry still serves. In all situations I can think of, the value of the **DigestMethod** entry is redundant at best, and unnecessarily confusing at worst.

Proposed solution: deprecate **DigestMethod**.

## Additional context

This discussion came up in an email conversation in the Encryption & Digital Signatures DG under WG8. It's relevant for ISO 32001: if **DigestMethod** still serves a useful purpose, ISO 32001 would probably have to update its range of possible values.

PS: I'm also not sure what the rationale for the note added in the 2020 edition was, esp. the reason why the entry was made required. If someone could enlighten me, that'd be great! :)

-------

EDIT: For completeness & ease of reference, here's the corresponding definition in ISO 32000-1:2008, Table 253:

> (Optional; PDF 1.5 required) A name identifying the algorithm that shall be used when computing the digest. Valid values are _MD5_ and _SHA1_. Default value: _MD5_. For security reasons, MD5 should not be used. It is mentioned for backwards compatibility, since it remains the default value.

**Comment by mkl-public** [comment]
> I'm also not sure what the rationale for the note added in the 2020 edition was, esp. the reason why the entry was made required. If someone could enlighten me, that'd be great! :)

To me it sounds like the person who added that sentence was not aware of the history of *DigestMethod* and assumed it was sensible somehow. On this supposition it makes sense to make the entry required after its default value has been declared deprecated: Not providing an explicit value would after all mean defaulting to something deprecated...

**Comment by lrosenthol** [comment]
A well known implementation already considers `DigestMethod` and `DigestValue` to be deprecated, by never writing them and only reading/using them when present.

**Comment by petervwyatt** [comment]
PDF TWG agree to flag this key as deprecated (in the ISO 32000-2 sense)

**Comment by petervwyatt** [comment]
If it's deprecated and shouldn't be written, then its also becoming optional in PDF 2.0.

**Comment by a20dev** [comment]
Some additional history explaining why some software still adds DigestMethod: Implementation note 144 of PDF Reference 1.7 reads

```
In Acrobat 6.0 and greater, the DigestMethod entry in the signature refer-
ence dictionary is required, even though this specification indicates that
entry is optional.
```
